torch/optim/asgd.py

```diff
@@ -246,7 +246,9 @@ def _single_tensor_asgd(
             param.add_(grad, alpha=-eta_value)  # update parameter
 
         # averaging
-        if capturable or mu.item() != 1:
+        # The compiler will only launch one kernel
+        # and it does not support data-dependent control flow
+        if torch._utils.is_compiling() or capturable or mu.item() != 1:
```

Comment:
is this is_compiling() necessary? my expectation is that we want to encapsulate the `is_compiling()`s into contained spots like `_get_value` that we can mentally keep track of and not broadstroke codepath changes.

Reply:
Yes it is, we can't use get_value here because we are actually basing control flow off of the value inside a tensor (rather than just running computation on it). In this case eager has an optimization where it only launches one kernel if the value is 1. For compile this optimization doesn't matter because we fuse the extra kernels into one kernel anyway. 

I could use torch.where but this basically has the same problem where we should only call it if we are compiling and use the item call for eager.